R-package/src/install.libs.R

```diff
@@ -12,7 +12,77 @@ R_ver <- as.double(R.Version()$major) + as.double(R.Version()$minor) / 10.0
 
 if (!(R_int_UUID == "0310d4b8-ccb1-4bb8-ba94-d36a55f60262"
     || R_int_UUID == "2fdf6c18-697a-4ba7-b8ef-11c0d92f1327")) {
-  print("Warning: unmatched R_INTERNALS_UUID, may cannot run normally.")
+  print("Warning: unmatched R_INTERNALS_UUID, may not run normally.")
+}
+
+# system() will not raise an R exception if the process called
+# fails. Wrapping it here to get that behavior.
+#
+# system() introduces a lot of overhead, at least on Windows,
+# so trying processx if it is available
+.run_shell_command <- function(cmd, args, strict = TRUE) {
+    on_windows <- .Platform$OS.type == "windows"
+    has_processx <- suppressMessages({
+      suppressWarnings({
+        require("processx")  # nolint
+      })
+    })
+    if (has_processx && on_windows) {
+      result <- processx::run(
+        command = cmd
+        , args = args
+        , windows_verbatim_args = TRUE
+        , error_on_status = FALSE
+        , echo = TRUE
+      )
+      exit_code <- result$status
+    } else {
+      if (on_windows) {
+        message(paste0(
+          "Using system() to run shell commands. Installing "
+          , "'processx' with install.packages('processx') might "
+          , "make this faster."
+        ))
+      }
+      cmd <- paste0(cmd, " ", paste0(args, collapse = " "))
+      exit_code <- system(cmd)
+    }
+
+    if (exit_code != 0L && isTRUE(strict)) {
+        stop(paste0("Command failed with exit code: ", exit_code))
+    }
+    return(invisible(exit_code))
+}
+
+# try to generate Visual Studio build files
+.generate_vs_makefiles <- function(cmake_args) {
+  vs_versions <- c(
+    "Visual Studio 16 2019"
+    , "Visual Studio 15 2017"
+    , "Visual Studio 14 2015"
+  )
+  working_vs_version <- NULL
+  for (vs_version in vs_versions) {
+    print(sprintf("Trying '%s'", vs_version))
+    # if the build directory is not empty, clean it
+    if (file.exists("CMakeCache.txt")) {
+      file.remove("CMakeCache.txt")
```

Comment:
Is it enough to remove only one cache file to reset CMake state?

Reply:
yes, I believe so. In the Azure build we use Visual Studio 15 2017 but the first things searched by this code is Visual Studio 16 2019, so I think it's working correctly.